server/validation/validation.go

```diff
@@ -1,47 +1,313 @@
+			Reason:   fmt.Sprintf("unexpected rewrite encountered with tupelset relation '%s#%s'", objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	user := tk.GetUser()
+
+	// if a tupleset relation is related to an object but not a typed wildcard (e.g. 'user:*')
+	// then it is valid
+	if tuple.IsValidObject(user) && !tuple.IsTypedWildcard(user) {
+		return nil
+	}
+
+	// tupleset relation involving a wildcard (covers the '*' and 'type:*' cases)
+	if tuple.IsWildcard(user) {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("unexpected wildcard relationship with tupleset relation '%s#%s'", objectType, relation),
+			TupleKey: tk,
+		}
+	}
+
+	// tupleset relation involving a userset (e.g. object#relation)
+	return &tuple.InvalidTupleError{
+		Reason:   fmt.Sprintf("unexpected userset relationship '%s' with tupleset relation '%s#%s'", user, objectType, relation),
+		TupleKey: tk,
+	}
+}
+
+// validateTypeRestrictions makes sure the type restrictions are enforced.
+// 1. If the tuple is of the form (user=person:bob, relation=reader, object=doc:budget), then the type "doc", relation "reader" must allow type "person".
+// 2. If the tuple is of the form (user=group:abc#member, relation=reader, object=doc:budget), then the type "doc", relation "reader" must allow type "group", relation "member".
+// 3. If the tuple is of the form (user=*, relation=reader, object=doc:budget), we allow it only if the type "doc" relation "reader" allows at least one type (with no relation)
+func validateTypeRestrictions(model *openfgapb.AuthorizationModel, tk *openfgapb.TupleKey) error {
+	objectType := tuple.GetType(tk.GetObject())           // e.g. "doc"
+	userType, userID := tuple.SplitObject(tk.GetUser())   // e.g. (person, bob) or (group, abc#member) or ("", *)
+	_, userRel := tuple.SplitObjectRelation(tk.GetUser()) // e.g. (person:bob, "") or (group:abc, member) or (*, "")
+
+	ts := typesystem.New(model)
+
+	typeDefinitionForObject, ok := ts.GetTypeDefinition(objectType)
+	if !ok {
+		return &tuple.InvalidTupleError{
+			Reason:   fmt.Sprintf("type '%s' does not exist in the authorization model", objectType),
+			TupleKey: tk,
+		}
+	}
+
+	relationsForObject := typeDefinitionForObject.GetMetadata().GetRelations()
+	if relationsForObject == nil {
+		if ts.GetSchemaVersion() == typesystem.SchemaVersion1_1 {
+			// if we get here, there's a bug in the validation of WriteAuthorizationModel API
+			msg := "invalid authorization model"
+			return fmt.Errorf(msg)
+		}
+
+		// authorization model is old/unspecified and does not have type information
+		return nil
+	}
+
+	// at this point we know the auth model has type information
+	if userType != "" {
+		if _, ok := ts.GetTypeDefinition(userType); !ok {
+			return &tuple.InvalidTupleError{
+				Reason:   fmt.Sprintf("undefined type for user '%s'", tk.GetUser()),
+				TupleKey: tk,
+			}
+		}
+	}
+
+	relationInformation := relationsForObject[tk.Relation]
+
+	// case 1
+	if userRel == "" && userID != "*" {
```

Comment:
This code needs some work now that we have typed and untyped wildcard behavior. This path seems to only account for the untyped wildcard behavior. 

- [ ] @jon-whit make sure the typed wildcard path is accounted for

Reply:
Follow up.